### PR TITLE
Replayable costs refactors

### PR DESCRIPTION
**Pull request by ArturGajowy**
## Overview
<sup>_What this PR does, and why it's needed_</sup>

Refactors and a minor change to charging I've gathered when working on making execution costs replayable.

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by ArturGajowy** [comment]
bors try

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala [hunk 1]
```diff
@@ -90,35 +90,37 @@ class CostAccountingSpec extends FlatSpec with Matchers with PropertyChecks with
...
+      "@1!(1) | @2!(2) | @3!(3)",
+      5L + parsingCost,
+      List(Cost(parsingCost, "parsing"), Cost(4, "substitution"), Cost(4, "substitution"))
+    )
+  }
+
+  private def checkPhloLimitExceeded(
+      contract: String,
+      initialPhlo: Long,
+      expectedCosts: List[Cost]
```
Comment (tymm) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala]: Nitpick: `Seq[Cost]`

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/ChargingRSpace.scala [hunk 2]
```diff
@@ -48,10 +46,8 @@ object ChargingRSpace {
         Option[(ContResult[Par, BindPattern, TaggedContinuation], Seq[Result[ListParWithRandom]])]
       ] =
         for {
-          _ <- charge[F](storageCostConsume(channels, patterns, continuation))
-          peekChannels = if (peek) {
-            SortedSet((0 to channels.size - 1): _*)
-          } else SortedSet.empty[Int]
+          _            <- charge[F](storageCostConsume(channels, patterns, continuation))
+          peekChannels = if (peek) SortedSet(channels.indices: _*) else SortedSet.empty[Int]
```
Comment (tymm) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/ChargingRSpace.scala:48]: Are you aware that this is a change in what the `SortedSet` looks like?
```
val c = Seq(1,2,3)
0 to c.size - 1 => Range 0 to 2
c.indices => Range 0 to 3
```

**Review by tymm — CHANGES_REQUESTED** [review]

**Review by tymm — CHANGES_REQUESTED** [review]
> If you've read this far, please let me know if I should've bothered
writing that. Thx!

Verbosity like this in commit messages is a blessing for me!

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/ChargingRSpace.scala (hunk 2 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/ChargingRSpace.scala:48]: From my repl:
```
val c = Seq(1,2,3)
0 to c.size - 1  
c.indices 
res0 == res1

//results

c: Seq[Int] = List(1, 2, 3)
res0: scala.collection.immutable.Range.Inclusive = Range 0 to 2
res1: scala.collection.immutable.Range = Range 0 until 3
res2: Boolean = true
```

So the ranges are equal, despite having different ends (one is inclusive) :)

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/ChargingRSpace.scala:48]: Thx for being vigilant! :+1:

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala (hunk 1 above)
Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingSpec.scala]: Done :)

**Review by tymm — APPROVED** [review]

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 3]
```diff
@@ -1046,8 +1046,8 @@ class DebruijnInterpreter[M[_], F[_]](
             else ().pure[M]
         baseExpr <- evalSingleExpr(p)
         element  <- evalExpr(args.head)
-        result   <- delete(baseExpr, element)
         _        <- charge[M](REMOVE_COST) //TODO(mateusz.gorski): think whether deletion of an element from the collection should dependent on the collection type/size
+        result   <- delete(baseExpr, element)
```
Comment (lukasz-golebiewski) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1050]: Wouldn't it make more sense to move the `charge` calls inside the functions they are charging for? Thanks to that such call won't be forgotten when one of those methods is called elsewhere

**Review by lukasz-golebiewski — APPROVED** [review]
LGTM

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 3 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala:1050]: That seems doable, will consider it the next time I break things in there :+1:

**Pull request merged**